src/app/PersistentData.h

```diff
@@ -0,0 +1,173 @@
+
+        uint8_t buffer[kMaxSerializedSize] = { 0 };
+        DefaultStorageKeyAllocator key;
+
+        // Set data to defaults
+        Clear();
+
+        // Update storage key
+        ReturnErrorOnFailure(UpdateKey(key));
+
+        // Load the serialized data
+        uint16_t size  = static_cast<uint16_t>(sizeof(buffer));
+        CHIP_ERROR err = storage->SyncGetKeyValue(key.KeyName(), buffer, size);
+        VerifyOrReturnError(CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND != err, CHIP_ERROR_NOT_FOUND);
+        ReturnErrorOnFailure(err);
+
+        // Decode serialized data
+        TLV::TLVReader reader;
+        reader.Init(buffer, size);
+        return Deserialize(reader);
+    }
+
+    virtual CHIP_ERROR Delete(PersistentStorageDelegate * storage)
+    {
+        VerifyOrReturnError(nullptr != storage, CHIP_ERROR_INVALID_ARGUMENT);
+
+        DefaultStorageKeyAllocator key;
+        // Update storage key
+        ReturnErrorOnFailure(UpdateKey(key));
+        // Delete stored data
+        return storage->SyncDeleteKeyValue(key.KeyName());
+    }
+};
+
+struct FabricList : public PersistentData<kPersistentBufferMax>
+{
+    static constexpr TLV::Tag TagFirstFabric() { return TLV::ContextTag(1); }
+    static constexpr TLV::Tag TagFabricCount() { return TLV::ContextTag(2); }
+
+    chip::FabricIndex first_fabric = kUndefinedFabricIndex;
+    uint8_t fabric_count           = 0;
+
+    FabricList() = default;
+    FabricList(chip::FabricIndex first) : first_fabric(first), fabric_count(1) {}
+
+    CHIP_ERROR UpdateKey(DefaultStorageKeyAllocator & key) override
+    {
+        key.FabricList();
+        return CHIP_NO_ERROR;
+    }
+
+    void Clear() override
+    {
+        first_fabric = kUndefinedFabricIndex;
+        fabric_count = 0;
+    }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        TLV::TLVType container;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, container));
+
+        ReturnErrorOnFailure(writer.Put(TagFirstFabric(), static_cast<uint16_t>(first_fabric)));
+        ReturnErrorOnFailure(writer.Put(TagFabricCount(), static_cast<uint16_t>(fabric_count)));
+
+        return writer.EndContainer(container);
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag()));
+        VerifyOrReturnError(TLV::kTLVType_Structure == reader.GetType(), CHIP_ERROR_INTERNAL);
+
+        TLV::TLVType container;
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        // first_fabric
+        ReturnErrorOnFailure(reader.Next(TagFirstFabric()));
+        ReturnErrorOnFailure(reader.Get(first_fabric));
+        // fabric_count
+        ReturnErrorOnFailure(reader.Next(TagFabricCount()));
+        ReturnErrorOnFailure(reader.Get(fabric_count));
+
+        return reader.ExitContainer(container);
+    }
+};
+
+/**
+ * Template used to iterate the stored data
```

Comment:
This template does not seem to relate directly to stored data.

Should we place things under a deeper namespace than just `chip` ?